### PR TITLE
Add nmstate handler daemonset to operator

### PR DESCRIPTION
**Pull request by sjpotter**
per my email to kubevirt-dev, here's my work in progress to add the nmstate handler from kubernetes-nmstate repo to the operator.

still have to check into the manifests it builds to see if anything has to be added, but figure get the PR out early so general comments can be made if I didn't understand something of how its supposed to work.

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

File: data/nmstate/001-rbac [hunk 1]
```diff
@@ -0,0 +1,64 @@
+{{ if .EnableSCC }}
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: We need the RBAC regardless if scc is enable.

but we need to add a section of scc 

you can take a look on this yaml https://github.com/kubevirt/cluster-network-addons-operator/blob/master/data/sriov/001-rbac.yaml

something like this
```
{{ if .EnableSCC }}
---
apiVersion: security.openshift.io/v1
kind: SecurityContextConstraints
metadata:
  name: nmstate
allowPrivilegedContainer: true
allowHostDirVolumePlugin: true
runAsUser:
  type: RunAsAny
seLinuxContext:
  type: RunAsAny
users:
- system:serviceaccount:nmstate:nmstate-controller
{{ end }}
```

File: data/nmstate/002-nmstated.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+metadata:
+  name: nm-state-controller
+  namespace: nmstate-handler
+spec:
+  template:
+    metadata:
+      labels:
+        nmstate.io: state-controller
+    spec:
+{{ if .EnableSCC }}
```
Comment (SchSeba) [inline data/nmstate/002-nmstated.yaml]: we need the user always not only if the scc is enable

File: data/nmstate/002-nmstated.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+---
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  name: nm-state-controller
+  namespace: nmstate-handler
```
Comment (SchSeba) [inline data/nmstate/002-nmstated.yaml]: please change the namespace here to just `nmstate`

File: data/nmstate/001-rbac [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+subjects:
+  - kind: ServiceAccount
+    name: nmstate-state-controller
+    namespace: nmstate-default
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: nmstate-state-controller
+  namespace: nmstate-default
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: please change the namespace here to just `nmstate`

File: data/nmstate/001-rbac [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  labels:
+    nmstate.io: ""
+roleRef:
+  kind: ClusterRole
+  name: nmstate-state-controller
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+  - kind: ServiceAccount
+    name: nmstate-state-controller
+    namespace: nmstate-default
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: please change the namespace here to just `nmstate`

File: data/nmstate/001-rbac [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  name: nmstate-state-controller
+  namespace: nmstate-default
+  labels:
+    nmstate.io: ""
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  name: nmstate-state-controller
+  namespace: nmstate-default
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: please change the namespace here to just `nmstate`

File: data/nmstate/001-rbac [hunk 7]
```diff
@@ -0,0 +1,64 @@
+{{ if .EnableSCC }}
+kind: ServiceAccount
+apiVersion: v1
+metadata:
+  name: nmstate-state-controller
+  namespace: nmstate-default
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: please change the namespace here to just `nmstate`

File: data/nmstate/001-rbac [hunk 8]
```diff
@@ -0,0 +1,64 @@
+{{ if .EnableSCC }}
+kind: ServiceAccount
+apiVersion: v1
+metadata:
+  name: nmstate-state-controller
```
Comment (SchSeba) [inline data/nmstate/001-rbac]: I think we can change it to `nmstate-controller`

File: data/nmstate/000-ns.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: nmstate-default
```
Comment (SchSeba) [inline data/nmstate/000-ns.yaml]: please change the namespace here to just nmstate

File: data/nmstate/002-nmstated.yaml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+spec:
+  template:
+    metadata:
+      labels:
+        nmstate.io: state-controller
+    spec:
+{{ if .EnableSCC }}
+      serviceAccountName: nmstate-state-controller
```
Comment (SchSeba) [inline data/nmstate/002-nmstated.yaml]: please change the service account to `nmstate-controller`

File: pkg/components/components.go [hunk 11]
```diff
@@ -19,6 +19,7 @@ const (
 	SriovDpImageDefault        = "quay.io/booxter/sriov-device-plugin:latest"
 	SriovCniImageDefault       = "docker.io/nfvpe/sriov-cni:latest"
 	KubeMacPoolImageDefault    = "quay.io/kubevirt/kubemacpool:v0.2.0"
+	NMStateHandlerImageDefault = "quay.io/nmstate/kubernetes-nmstate-state-handler:latest"
```
Comment (SchSeba) [inline pkg/components/components.go]: Can we tag the nmstate image with a release and change the latest tag here?

File: pkg/network/nmstate.go [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+func changeSafeNMState(prev, next *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if prev.NMState != nil && !reflect.DeepEqual(prev.NMState, next.NMState) {
+		return []error{errors.Errorf("cannot modify NMState Daemon configuration once it is deployed")}
+	}
+	return nil
+}
+
+// renderNMState generates the manifests of NMState handler
+func renderNMState(conf *opv1alpha1.NetworkAddonsConfigSpec, manifestDir string, clusterInfo *ClusterInfo) ([]*unstructured.Unstructured, error) {
+	if conf.LinuxBridge == nil {
```
Comment (SchSeba) [inline pkg/network/nmstate.go]: copy/paste issue? should be `conf.NMState`

File: pkg/network/nmstate.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["NMStateHandlerImage"] = os.Getenv("NMSTATE_HANDLER_IMAGE")
```
Comment (SchSeba) [inline pkg/network/nmstate.go]: maybe just `NMSTATE_IMAGE` without the `HANDLER` ?

File: data/nmstate/001-rbac (hunk 8 above)
Comment (phoracek) [reply data/nmstate/001-rbac]: I think we should keep it nmstate-state-controller. we will introduce nmstate-policy-controller once it is ready

File: pkg/components/components.go (hunk 11 above)
Comment (phoracek) [reply pkg/components/components.go]: I will do the release later, we can keep it like this for now. I don't want the PR to be blocked on me

File: pkg/network/nmstate.go (hunk 13 above)
Comment (phoracek) [reply pkg/network/nmstate.go]: actually this data and environment variable should be nmstate state handler image

File: data/nmstate/001-rbac (hunk 1 above)
Comment (sjpotter) [reply data/nmstate/001-rbac]: ok, understood.  SCC is an openshift construct that one can use to control how things are being executed (in this case as a service account)

File: pkg/network/nmstate.go (hunk 12 above)
Comment (sjpotter) [reply pkg/network/nmstate.go]: yes, woops.  will fix.

File: data/nmstate/001-rbac (hunk 8 above)
Comment (sjpotter) [reply data/nmstate/001-rbac]: any conclusion on what's preferred?

Comment (SchSeba) [reply data/nmstate/001-rbac]: Oh I did remember we have another component call policy so yes please leave it like that.

File: data/nmstate/002-nmstated.yaml (hunk 10 above)
Comment (sjpotter) [reply data/nmstate/002-nmstated.yaml]: per discussion above, keeping this the same

File: pkg/network/nmstate.go (hunk 13 above)
Comment (sjpotter) [reply pkg/network/nmstate.go]: so this becomes wordy, at least as far as I understand the comment,  we end up with something like NMStateState in multiple places.  could we just abbreviate it for the future of having nmstate and nmpolicy?  I'll update the PR per my understanding of the comment.

Comment (phoracek) [reply pkg/network/nmstate.go]: It is a little too wordy, but I prefer that over implicit names. It's hidden from user anyway.

File: data/nmstate/002-nmstated.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        beta.kubernetes.io/arch: amd64
+      tolerations:
+        - key: node-role.kubernetes.io/master
+          operator: Exists
+          effect: NoSchedule
+      containers:
+        - name: state-controller
+          image: {{ .NMStateHandlerImage }}
+          imagePullPolicy: {{ .ImagePullPolicy }}
+          args: ["-execution-type", "client"]
```
Comment (phoracek) [inline data/nmstate/002-nmstated.yaml]: I think we have a bug in kubernetes-nmstate manifests. This type should be controller, not client. I'm sorry for dragging this PR for so long, but would you please fix it here?

File: pkg/network/nmstate.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+// renderNMState generates the manifests of NMState handler
+func renderNMState(conf *opv1alpha1.NetworkAddonsConfigSpec, manifestDir string, clusterInfo *ClusterInfo) ([]*unstructured.Unstructured, error) {
+	if conf.NMState == nil {
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["NMStateStateHandlerImage"] = os.Getenv("NMSTATE_STATEHANDLER_IMAGE")
```
Comment (phoracek) [inline pkg/network/nmstate.go]: NMSTATE_STATE_HANDLER_IMAGE

Comment (sjpotter) [reply pkg/network/nmstate.go]: also then changed in pkg/components/components.go

File: README.md [hunk 16]
```diff
@@ -136,6 +137,14 @@ spec:
    rangeEnd: "FD:FF:FF:FF:FF:FF"
 ```
 
+## Nmstate
+
+The operator allows the administrator to deploy the [NmState State Controller](https://github.com/nmstate/nmstate) as a daemonset across all of one's nodes.
+This project manages host networking settings in a declarative manner. The networking state is described by a pre-defined schema. Reporting of current state and changes to it (desired state) both conform to the schema.
+Nmstate is aimed to satisfy enterprise needs to manage host networking through a northbound declarative API and multi provider support on the southbound. NetworkManager acts as the main (and currently the only) provider supported.
+
+Future work will include managing Nmstate settings through the operator as well
```
Comment (SchSeba) [inline README.md]: Can you start with `**Note:** ` and add a point at the end please

Comment (phoracek) [reply README.md]: Or maybe remove the note? There is no need to describe future work here IMO, better open an issue to track the effort. Just an idea, won't block with this.

File: cluster/functest.sh [hunk 17]
```diff
@@ -51,6 +51,7 @@ spec:
   linuxBridge: {}
   multus: {}
   sriov: {}
+  nmState: {}
```
Comment (phoracek) [inline cluster/functest.sh]: does not match lowercase naming used in REAEDME

File: pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go [hunk 18]
```diff
@@ -13,6 +13,7 @@ type NetworkAddonsConfigSpec struct {
 	Sriov           *Sriov            `json:"sriov,omitempty"`
 	KubeMacPool     *KubeMacPool      `json:"kubeMacPool,omitempty"`
 	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	NMState         *NMState           `json:"nmState,omitempty"`
```
Comment (phoracek) [inline pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go]: please switch to the lowercase `nmstate`

**Comment by sjpotter** [comment]
do you guys want a form of history preserved in the squash or just what's basically the first message.

**Comment by phoracek** [comment]
@sjpotter just the first one

**Comment by sjpotter** [comment]
can't seem to figure out why github still has "changes requested".

**Comment by SchSeba** [comment]
@sjpotter sorry but we merge a PR can you please rebase on master and fix the conflict

**Comment by SchSeba** [comment]
@sjpotter I still see a conflict problem can you please fix it?

**Comment by sjpotter** [comment]
I've been away this weekend and today can get to it tomorrow.

**Comment by sjpotter** [comment]
the above / continued rebase issue is due to whitespace, so every time a new version of an image is comitted, I have to rebase, as my PR changes whitespace in all the images.  I could insert a blank line in the consts, to avoid gofmt adding whitespace.

**Comment by phoracek** [comment]
@sjpotter sorry about that. Let's wait for CI to pass and then I will merge.

**Comment by phoracek** [comment]
ci test please

File: pkg/network/nmstate.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+// renderNMState generates the manifests of NMState handler
+func renderNMState(conf *opv1alpha1.NetworkAddonsConfigSpec, manifestDir string, clusterInfo *ClusterInfo) ([]*unstructured.Unstructured, error) {
+	if conf.NMState == nil {
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["NMStateStateHandlerImage"] = os.Getenv("NMSTATE_STATE_HANDLER_IMAGE")
```
Comment (SchSeba) [inline pkg/network/nmstate.go]: related to CI error 
```
status:
  conditions:
  - lastProbeTime: 2019-05-09T20:06:57Z
    lastTransitionTime: 2019-05-09T20:06:10Z
    message: 'failed to render: failed to render nmstate state handler manifests:
      error rendering manifests: failed to render manifest data/nmstate/002-nmstated.yaml:
      template: data/nmstate/002-nmstated.yaml:22:20: executing "data/nmstate/002-nmstated.yaml"
      at <.NMStateHandlerImage>: map has no entry for key "NMStateHandlerImage"'
    reason: FailedToRender
    status: "True"
    type: Failing
FAILED
```

can you please change here to 
```
NMStateStateHandlerImage -> NMStateHandlerImage
```

Comment (sjpotter) [reply pkg/network/nmstate.go]: I think we said we wanted it to be verbosely named?  unless the same image will be used for both controller and client?

Comment (sjpotter) [reply pkg/network/nmstate.go]: I've changed the template to be using the correct variable as defined in pkg/network/nmstate.go

**Comment by SchSeba** [comment]
ci tests please

**Comment by SchSeba** [comment]
ci test please
